.circleci/config.yml

```diff
@@ -45,15 +45,67 @@ jobs:
       - copy-gradle-properties
       - android/restore-gradle-cache
       - run:
-          name: Build
-          command: ./gradlew --stacktrace assembleDebug assembleRelease
+          name: Bintray Upload
+          command: |
+            # Use a different versioning style for builds from PRs to allow
+            # developers to iterate faster.
+            #
+            # All those dev builds will be deleted once the PR is merged by
+            # https://github.com/Automattic/bintray-garbage-collector/
+            #
+            # If $PREFIX is not set, that is, if we're not on a PR but a merge
+            # commit on develop or trunk, the version will be the value from
+            # the source code.
+            branch=$CIRCLE_BRANCH
+            if [[ -n "$CIRCLE_PULL_REQUEST" ]]; then
+              PR_NUMBER=$(basename $CIRCLE_PULL_REQUEST)
+              # Making the assumption that $CIRCLE_SHA1 will always be
+              # available.
+              PREFIX="$PR_NUMBER-$CIRCLE_SHA1"
+            elif [[ "$branch" != "trunk" && "$branch" != "develop" ]]; then
+              # This happens on the first push of a new branch, when there
+              # isn't a PR open for it yet.
+              echo "Running on a feature branch with no open PR: skipping Bintray upload"
+              exit 0
+            fi
+
+            # Locally, calling `./gradlew bintrayUpload` is enough to generate
+            # all the required artifacts. On CI, unless run each task
+            # individually, the Bintray task will fail to find the AAR and
+            # POM artifacts in the build/ folder.
+            #
+            # This builds the AAR.
+            ./gradlew --stacktrace assembleRelease
+
+            # This builds the POM.
+            #
+            # On local, the POM is generated automatically as part of
+            # assembleRelease, but that doesn't seem to be the case on CI. Here
+            # we need to explicitly generated it, otherwise bintrayUpload won't
+            # find it and won't upload it.
+            #
+            # Worth noting that the Bintray plugin mentions the necessity for
+            # a workaround to generate the POM. See:
+            # https://github.com/bintray/gradle-bintray-plugin#maven-publications
+            #
+            # See also the description and comments in this PR:
+            # https://github.com/wordpress-mobile/WordPress-Utils-Android/pull/39
+            #
+            # Notice the `-PbintrayVersion` parameter to make sure the POM has
+            # the same version as the one bintrayUpload will use.
+            ./gradlew --stacktrace generatePomFileForUtilsPublicationPublication -PbintrayVersion=$PREFIX
+
+            # Finally, this uploads both AAR and POM to Bintray.
+            ./gradlew --stacktrace bintrayUpload -PbintrayVersion=$PREFIX
```

Comment:
My understanding of Gradle is pretty superficial. If you know why a task would behave so differently between CI and local, and have a suggestion on how to avoid this long winded workaround, I'd love to know about it.

The commits in #40 show the different things I tried before coming to this point, including builds failing without finding the artifact.

Reply:
This is weird. I wonder if it can be related to the different OS (OSX vs Linux). 
Anyway, I don't dislike that we have full control on what and how we build before uploading it tbh.

Reply:
That is really weird that the task dependencies would be different on CI. 🤔  It might be interesting to see what the output of `./gradlew bintrayUpload --dry-run` looks like on CI compared to what that command outputs locally.

Reply:
@mchowning good question 🤔 

I tried this out [in this build](https://app.circleci.com/pipelines/github/wordpress-mobile/WordPress-Utils-Android/150/workflows/a54c8976-68ee-41b8-8446-6bccdc89c7d2/jobs/298) via #52.

### CI output

```
$ ./gradlew bintrayUpload --dry-run -PbintrayVersion=test-dry-run-1

Starting a Gradle Daemon, 1 incompatible and 1 stopped Daemons could not be reused, use --status for details
Configuration on demand is an incubating feature.

> Configure project :WordPressUtils
WARNING: API 'variant.getJavaCompile()' is obsolete and has been replaced with 'variant.getJavaCompileProvider()'.
It will be removed in version 5.0 of the Android Gradle plugin.
For more information, see https://d.android.com/r/tools/task-configuration-avoidance.
To determine what is calling variant.getJavaCompile(), use -Pandroid.debug.obsoleteApi=true on the command line to display more information.
:WordPressUtils:bintrayUpload SKIPPED
:bintrayUpload SKIPPED
:bintrayPublish SKIPPED

BUILD SUCCESSFUL in 6s
```

### Local output

```
➜ ./gradlew bintrayUpload --dry-run -PbintrayVersion=test-dry-run-1  

> Configure project :WordPressUtils
WARNING: API 'variant.getJavaCompile()' is obsolete and has been replaced with 'variant.getJavaCompileProvider()'.
It will be removed in version 5.0 of the Android Gradle plugin.
For more information, see https://d.android.com/r/tools/task-configuration-avoidance.
To determine what is calling variant.getJavaCompile(), use -Pandroid.debug.obsoleteApi=true on the command line to display more information.
:WordPressUtils:preBuild SKIPPED
:WordPressUtils:preReleaseBuild SKIPPED
:WordPressUtils:compileReleaseAidl SKIPPED
:WordPressUtils:mergeReleaseJniLibFolders SKIPPED
:WordPressUtils:mergeReleaseNativeLibs SKIPPED
:WordPressUtils:stripReleaseDebugSymbols SKIPPED
:WordPressUtils:copyReleaseJniLibsProjectAndLocalJars SKIPPED
:WordPressUtils:compileReleaseRenderscript SKIPPED
:WordPressUtils:generateReleaseBuildConfig SKIPPED
:WordPressUtils:generateReleaseResValues SKIPPED
:WordPressUtils:generateReleaseResources SKIPPED
:WordPressUtils:packageReleaseResources SKIPPED
:WordPressUtils:parseReleaseLocalResources SKIPPED
:WordPressUtils:processReleaseManifest SKIPPED
:WordPressUtils:generateReleaseRFile SKIPPED
:WordPressUtils:compileReleaseKotlin SKIPPED
:WordPressUtils:javaPreCompileRelease SKIPPED
:WordPressUtils:compileReleaseJavaWithJavac SKIPPED
:WordPressUtils:extractReleaseAnnotations SKIPPED
:WordPressUtils:mergeReleaseGeneratedProguardFiles SKIPPED
:WordPressUtils:mergeReleaseConsumerProguardFiles SKIPPED
:WordPressUtils:mergeReleaseShaders SKIPPED
:WordPressUtils:compileReleaseShaders SKIPPED
:WordPressUtils:generateReleaseAssets SKIPPED
:WordPressUtils:packageReleaseAssets SKIPPED
:WordPressUtils:packageReleaseRenderscript SKIPPED
:WordPressUtils:prepareLintJarForPublish SKIPPED
:WordPressUtils:processReleaseJavaRes SKIPPED
:WordPressUtils:mergeReleaseJavaResource SKIPPED
:WordPressUtils:syncReleaseLibJars SKIPPED
:WordPressUtils:bundleReleaseAar SKIPPED
:WordPressUtils:generateMetadataFileForUtilsPublicationPublication SKIPPED
:WordPressUtils:generatePomFileForUtilsPublicationPublication SKIPPED
:WordPressUtils:publishUtilsPublicationPublicationToMavenLocal SKIPPED
:WordPressUtils:bintrayUpload SKIPPED
:bintrayUpload SKIPPED
:bintrayPublish SKIPPED

BUILD SUCCESSFUL in 743ms
```

There's quite a difference 😳

Reply:
> ...There's quite a difference 😳

For sure! That is so weird. I do not have a good idea why this wouldn't be working. Maybe the `WordPressUtils:publishUtilsPublicationPublicationToMavenLocal` doesn't get created on CI for some reason and that somehow breaks the task dependency chain? That's really just wild speculation on my part to be honest though. I'm a bit stumped.

You may have already tried this, but I'm curious what happens if you explicitly make the `bintrayPublish` task depend on `generatePomFileForUtilsPublicationPublication`, and then, if necessary, also explicitly make `generatePomFileForUtilsPublicationPublication` depend on `assembleRelease`? It doesn't seem like we should _have_ to do that, but it might fix the problem.

FWIW, what you have here seems fine, so I wouldn't disagree if you felt like it wasn't worth spending more time figuring this out. 

Even if we have to be explicit about invoking each task, It would be nice if we could combine them into a single build with something like `./gradlew assembleRelease generatePomFileForUtilsPublicationPublication bintrayUpload`. That would speed up the builds by not re-doing work (although caching reduces the re-work, I believe that at least the configuration phase will get re-run each time). BUT, given that the task dependencies clearly aren't working the way we expect on CI, I'm not feeling very confident that the tasks would always be run in the right order if we combined them into one build command. 😞